docs/components/query-frontend.md

```diff
@@ -0,0 +1,128 @@
+---
+title: Query Frontend
+type: docs
+menu: components
+---
+
+# Query Frontend
+
+The `thanos query-frontend` command implements a service that can be put in front of Thanos Queriers to improve the read path. It is
+based on the [Cortex Query Frontend](https://cortexmetrics.io/docs/architecture/#query-frontend) component so you can find some common features like `Splitting` and `Results Caching`.
+
+Query Frontend is fully stateless and horizontally scalable.
+
+Example command to run Query Frontend:
+
+```bash
+thanos query-frontend \
+    --http-address     "http://0.0.0.0:9090" \
+    --query-frontend.downstream-url="<thanos-querier>:<querier-http-port>"
+```
+
+_**NOTE:** Currently only range queries (`/api/v1/query_range` API call) are actually processed through Query Frontend. All other
+API calls just directly go to the downstream Querier, which means only range queries are split and cached. But we are planning to support instant queries as well.
+
+For more information please check out [initial design proposal](https://thanos.io/proposals/202004_embedd_cortex_frontend.md/).
+
+## Features
+
+### Splitting
+
+Query Frontend splits a long query into multiple short queries based on the configured `--query-range.split-interval` flag. The default value of `--query-range.split-interval`
+is `24h`. Set it to `0` disables query splitting, but please note that caching is also disabled in this case.
+
+There are some benefits from query splitting:
+
+1. Safe guard. It prevents large queries from causing OOM issues to Queries.
+2. Better parallelization.
+3. Better load balancing for Queries.
+
+### Retry
+
+Query Frontend supports a retry mechanism to retry query when HTTP requests are failing. There is a `--query-range.max-retries-per-request` flag to limit the maximum retry times.
+
+### Caching
+
+Query Frontend supports caching query results and reuses them on subsequent queries. If the cached results are incomplete,
+Query Frontend calculates the required subqueries and executes them in parallel on downstream queriers. Query Frontend can optionally align queries with their step parameter to improve the cacheability of the query results.
+
+Currently, only in-memory cache (fifo cache) is supported. An example config:
+
+```yaml
```

Comment:
this can be autogenerated. Can we do this in this PR maybe? (:

Reply:
I didn't know we can auto-generate this before. Let me take a look how to do it.

Reply:
yup, looks on object storage page (:

Reply:
When investigating this, I found actually some of our embedmd configs are wrong like `[embedmd]: # "../flags/config_index_cache_in_memory.txt yaml"`, this syntax is not supported. So I also changed them in that commit

Reply:
Thanks!